File: rulesets.js
import {config} from './config.js'
import { World  } from './models/world.js'

class BaseRuleset {
  constructor(policy, deckBuilder) {
    this.deckBuilder = deckBuilder
    this.policy = policy
    this.heroDeck = []
    this.bossDeck = []
  }

  makeWorld = function() {
    var hero = this.makeHero()
    var boss = this.makeBoss()
    if(this.heroDeck.length == 0){
      this.heroDeck = this.deckBuilder.buildFullDeck(this.deckBuilder.cards, config.DECK_SIZE)
    }
    if(this.bossDeck.length == 0){
      this.bossDeck = this.deckBuilder.buildFullDeck(this.deckBuilder.cards, config.DECK_SIZE)
    }
    this.world = new World([...this.deckBuilder.cards], hero, boss, [...this.heroDeck], [...this.bossDeck])
    this.world.isPlayerTurn = true
    return this.world;
  }

  makeHero = function() {
    var hero = this.deckBuilder.makeHero()
    hero.ruleset = this;
    return hero
  }

  makeBoss = function() {
    var boss = this.deckBuilder.makeBoss()
    boss.ruleset = this;
    return boss
  }

  removeCardsFromHand =  function(cards, hand) {
    for (var i = 0; i < cards.length; i++) {
      var card = cards[i]
      var index = hand.indexOf(card)
      if (index != -1) {
        hand.splice(index, 1)
      }
    }
    return hand
  }
  
  fillHand = function(hand, deck) {
    while(hand.length < config.HAND_SIZE) {
      hand.push(deck.pop())
    }
  }

  ensureHeroDeck = function(deck) {
    if (deck.length < config.DECK_SIZE) { 
      deck = this.deckBuilder.buildFullDeck(this.deckBuilder.cards, config.DECK_SIZE)
    }
    return deck
  }
}

class DamageOnlyRuleset extends BaseRuleset{
    constructor(policy, deckBuilder) {
        super(policy,deckBuilder)
    }
}

class DamageAndCostRuleset extends BaseRuleset{
    constructor(policy, deckBuilder) {
        super(policy, deckBuilder)
    }
}

export {DamageOnlyRuleset, DamageAndCostRuleset}

File: python/rulesets.py
from config import config
from models.world import World

class BaseRuleset:
    def __init__(self, policy, deckBuilder):
        self.deckBuilder = deckBuilder
        self.policy = policy
        self.heroDeck = []
        self.bossDeck = []

    def makeWorld(self):
        hero = self.makeHero()
        boss = self.makeBoss()
        if len(self.heroDeck) == 0:
            self.heroDeck = self.deckBuilder.buildFullDeck(self.deckBuilder.cards, config.DECK_SIZE)
        if len(self.bossDeck) == 0:
            self.bossDeck = self.deckBuilder.buildFullDeck(self.deckBuilder.cards, config.DECK_SIZE)
        self.world = World(self.deckBuilder.cards, hero, boss, self.heroDeck, self.bossDeck)
        self.world.isPlayerTurn = True
        return self.world

    def makeHero(self):
        hero = self.deckBuilder.makeHero()
        hero.ruleset = self
        return hero

    def makeBoss(self):
        boss = self.deckBuilder.makeBoss()
        boss.ruleset = self
        return boss

    def removeCardsFromHand(self, cards, hand):
        for card in cards:
            index = hand.index(card)
            if index != -1:
                hand.pop(index)
        return hand

    def fillHand(self, hand, deck):
        while len(hand) < config.HAND_SIZE:
            hand.append(deck.pop())

    def ensureHeroDeck(self, deck):
        if len(deck) < config.DECK_SIZE:
            deck = self.deckBuilder.buildFullDeck(self.deckBuilder.cards, config.DECK_SIZE)
        return deck



